### PR TITLE
Add high order gradient support for activation function

### PR DESCRIPTION
**Pull request by caogang**
[WIP] Add high order gradient support for sigmoid function, solving the issue #1483 

- [x] sigmoid

### CODE REVIEW

File: torch/autograd/_functions/pointwise.py [hunk 1]
```diff
@@ -61,22 +62,25 @@ def backward(self, grad_output):
...
-        backend = type2backend[type(result)]
-        backend.Sigmoid_updateGradInput(backend.library_state, None, grad_output,
-                                        grad_input, result)
-        return grad_input
+    @staticmethod
+    def backward(ctx, grad_output):
+        result, = ctx.saved_variables
+        grad_input = Variable(grad_output.data.new(ctx.input_size).zero_())
+        backend = type2backend[type(result.data)]
+        backend.Sigmoid_updateGradInput(backend.library_state, None, grad_output.data,
```
Comment (fmassa) [inline torch/autograd/_functions/pointwise.py]: This is not right,as you have unwrapped the data from the variable.
I think you should write the equation that was previous to this PR https://github.com/pytorch/pytorch/pull/1399 , but adding a case for volatile variables to make it more efficient.

Comment (caogang) [inline torch/autograd/_functions/pointwise.py]: Thks, I have update my code. However, I don't understand why it should use the equation  that was previous to this PR  #1399 when the `grad_output is not volatile`.

Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: Re-packing the grad in a new Variable doesn't retain the saved history so it's wrong. When a grad is volatile, then it doesn't have any history anyway, so this faster method is valid in this case.

File: torch/autograd/_functions/pointwise.py [hunk 2]
```diff
@@ -61,22 +62,28 @@ def backward(self, grad_output):
...
-        grad_input = grad_output.new()
-        backend = type2backend[type(result)]
-        backend.Sigmoid_updateGradInput(backend.library_state, None, grad_output,
-                                        grad_input, result)
-        return grad_input
+    @staticmethod
+    def backward(ctx, grad_output):
+        result, = ctx.saved_variables
+        if grad_output.volatile:
+            grad_input = Variable(grad_output.data.new(ctx.input_size).zero_())
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: you need to add `volatile=True` here

File: torch/autograd/_functions/pointwise.py (hunk 1 above)
Comment (caogang) [inline torch/autograd/_functions/pointwise.py]: oh, I know. Thank you for your explanation.

File: torch/autograd/_functions/pointwise.py (hunk 2 above)
Comment (caogang) [inline torch/autograd/_functions/pointwise.py]: ok, I will add it right now.

File: torch/autograd/_functions/pointwise.py [hunk 3]
```diff
@@ -61,22 +62,28 @@ def backward(self, grad_output):
 
 class Sigmoid(InplaceFunction):
 
-    def forward(self, i):
-        if self.inplace:
-            self.mark_dirty(i)
+    @staticmethod
+    def forward(ctx, i, inplace=False):
+        ctx.input_size = i.size()
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: Sorry, I missed this. It's unnecessary, you can just use `grad_output.size()` in `backward`.

File: torch/autograd/_functions/pointwise.py [hunk 4]
```diff
@@ -61,22 +62,28 @@ def backward(self, grad_output):
...
-        grad_input = grad_output.new()
-        backend = type2backend[type(result)]
-        backend.Sigmoid_updateGradInput(backend.library_state, None, grad_output,
-                                        grad_input, result)
-        return grad_input
+    @staticmethod
+    def backward(ctx, grad_output):
+        result, = ctx.saved_variables
+        if grad_output.volatile:
+            grad_input = Variable(grad_output.data.new(ctx.input_size).zero_(), volatile=True)
```
Comment (apaszke) [inline torch/autograd/_functions/pointwise.py]: Also, no need to call `zero_()` here (`updateGradInput` overwrites the tensor)

**Review by apaszke — APPROVED** [review]

**Comment by fmassa** [comment]
Are there tests for double backprop that could be easily added?

**Comment by apaszke** [comment]
per-op double backprop tests are unnecessary. We only use first-order jacobian vectror product functions to compute grads of any order, so as long as first-order is correct it should be all good (assuming autograd code is correct, but we have separate tests for that).

**Comment by fmassa** [comment]
That's not true for code that has a different behavior if the grad is volatile, as in this PR.

**Comment by soumith** [comment]
i think we should add gradgradcheck, just like gradcheck. We dont know if new-style functions have been written correctly for grad of grad out of the box (for example, user may have rewrapped a Variable somewhere and thought it was okay)

**Comment by apaszke** [comment]
`gradcheck` runs tests only on volatile grads, so this case is covered

**Comment by apaszke** [comment]
I think that instead of computing a full hessian of each op (these tests are going to be soooo slooow) we could just add some simple clauses to gradcheck that make sure that there exists a path from `grad_input.grad_fn` to `grad_output`'s grad accumulator. This should be enough.
